draft-bbs-signatures.md

```diff
@@ -746,24 +755,87 @@ Inputs:
 
 Parameters:
 
-- r (REQUIRED), non-negative integer. The prime order of the G_1 and G_2 groups,
-     defined by the ciphersuite.
+- r (REQUIRED), non-negative integer. The prime order of the G1 and G2 groups, defined by the ciphersuite.
+- expand_length (REQUIRED), non-negative integer. The number of bytes to read from the xof function, defined by the ciphersuite.
+- dst (REQUIRED), octet string. Domain separation tag.
 
 Outputs:
 
 - (scalar_1, ..., scalar_n), a list of non-zero scalars mod r.
 
 Procedure:
 
-1. h = xof(msg_octets)
+1. dst_prime = (I2OSP(len(dst), 1) || dst)
+
+2. msg_prime = (I2OSP(len(msg), 8) || msg)
+
+3. h = xof(msg_prime || dst_prime || I2OSP(0, 1) || I2OSP(n, 8))
+
+4. for i in (1, ..., n):
+
+5.     scalar_i = OS2IP(h.read(expand_length)) mod r
+
+6.     if scalar_i is 0, go back to step 3
+
+7. return (scalar_1, ..., scalar_n)
+```
+
+#### hash_to_scalar_xmd
+
+Implementations not wishing to use the more performant `hash_to_scalar_xof` operation, can elect to use `hash_to_scalar_xmd`. The `hash_to_scalar_xmd` is based on the `expand_message_xmd` function defined in Section 5.4 of [@!I-D.irtf-cfrg-hash-to-curve], with the addition of checking if the resulting scalar is 0.
+
+```
+result = hash_to_scalar_xmd(msg_octets, n)
+
+Inputs:
+- msg_octets (REQUIRED), octet string. The message to be hashed.
+- n (REQUIRED), non-negative integer. The number of scalars to output.
+
+Parameters:
+
+- r (REQUIRED), non-negative integer. The prime order of the G1 and G2 groups, defined by the ciphersuite.
+- expand_length (REQUIRED), non-negative integer. The number of bytes required to compute each scalar, defined by the ciphersuite.
+- b_in_bytes (REQUIRED), b / 8 for b the output size of H in bits. For example, for b = 256, b_in_bytes = 32.
+- s_in_bytes (REQUIRED), the input block size of H, measured in bytes. For example, for SHA-256, s_in_bytes = 64.
+- dst (REQUIRED), octet string. Domain separation tag.
+
+Procedure:
+
+1. ell = ceil(expand_length / b_in_bytes)
+
+2. dst_prime = dst || I2OSP(len(dst), 1)
```

Comment:
I think the DST length should be prepended for consistency

Reply:
It's true that this hurts consistency but there is a security reason for that. See [here](https://github.com/cfrg/draft-irtf-cfrg-hash-to-curve/issues/236) and [here](https://github.com/cfrg/draft-irtf-cfrg-hash-to-curve/pull/241) for context.

### TL;TR
 prepending len(DST) here will mean that there are different `msg`, `dst`, `expand_length` combinations that give the same hash output. 
 
### Note 
AFAIK there are 2 solutions to the above.

1. prepend the length of the message as well.
2. append the length of the dst only.

Neither are 100% consistent with the [Encoding of elements to be hashed](https://identity.foundation/bbs-signature/draft-bbs-signatures.html#name-encoding-of-elements-to-be-) section (for option 1 it may not be as "obvious", but after "appending" the lengths of some of the to-be-hashed elements one may thing that we would not need to append the length of the "entire" concatenated result as well).

For that reason, I prefer the current approach (option 2, i also update `hash_to_scalar_xof` to follow this). The "elements to be hashed" section will need updating IMO to something like "concatenating elements" or "Inputs to hash_to_scalar encoding".

Interested to hear others thoughts here as well.

Reply:
Im in favour of option 2 also